src/connector/Cargo.toml

```diff
@@ -106,7 +106,7 @@ risingwave_rpc_client = { workspace = true }
 rust_decimal = "1"
 serde = { version = "1", features = ["derive", "rc"] }
 serde_derive = "1"
-serde_json = "1"
+serde_json = { version = "1.0", features = ["preserve_order"] }
```

Comment:
This might be the reason for json encoder test to fail. Can you explain why `preserve_order` is required here?

Reply:
We need this feature to preserve the order of the keys as for the `serde_json::Map` in `recursive_parse_json`.

Reply:
Why should we keep the key order of json consistent with the order of sub-fields inserted into serde_json::Map?

Reply:
Why would you not to try keeping the original order? In my perspective this is like the column order in a traditional table.

Reply:
Firstly, the data we insert is a kv pair, so we do not need to preserve the order of insertion; Secondly, I think Any will eventually be a field on the top level because it is a `Jsonb` but not a `Struct`, so the sub-fields would not be columns but text.

Reply:
> Firstly, the data we insert is kv pairs, so we do not need to preserve the order of insertion; Secondly, I think Any will eventually be a field on the top level because it is a Jsonb but not a Struct, so the sub-fields would not be columns but text.

1. Everything user stores/retrieves from RW (or more general) could be generalized as a "kv pair", this is not the essence of this and also not the reason why we not preserving the order.

2. The sub_fields will have their types and will not be text, since user could further utilize them in RW, unless we could not make the transformation. See the example below.
![image](https://github.com/risingwavelabs/risingwave/assets/91381303/a192538e-18a3-4c1b-b091-a5f2408de4f4)

Reply:
> Everything user stores/retrieves from RW (or more general) could be generalized as a "kv pair", this is not the essence of this and also not the reason why we not preserving the order.

Since you can store/retrieves data with key, the order is not necessary. I think it's like the difference between `ordered_map` and `unordered_map`.

> The sub-fields will have their types and will not be text, since user could further utilize them in RW, unless we could not make the transformation. See the example below. (This is the example with full_name btw)

If Any and its sub-fields are useful in the downstream, it would be transformed into `Struct` but not `Jsonb`. Since we treat `Any` as `Jsonb` here, I suppose we do not utilize the structural information in RW, but pass it through to user. I'm not informed of the full background of this pr, so it's just a personal understanding. PTAL @tabVersion.

Reply:
> Since you can store/retrieves data with key, the order is not necessary.

We could store/retrieve data from table by specifying the key (column name here), if order is not necessary then why do we need `order by` support for this general purpose? Also it would be strange for user to see their input any messages displayed in a different order.

> If Any and its sub-fields are useful in the downstream, it would be transformed into Struct but not Jsonb. 

<img width="1426" alt="CleanShot 2023-10-25 at 23 32 14@2x" src="https://github.com/risingwavelabs/risingwave/assets/91381303/12e56f1c-94a4-4ea7-af15-26a9728e76ff">

Reply:
Making a clarification. 
`Any` is a placeholder for messages, it will not change the table schema, so we make it a `jsonb` field. And what we do here is extracting the message contents with the original message arch without wrapping.

> The sub-fields will have their types and will not be text, since user could further utilize them in RW, unless we could not make the transformation. See the example below. (This is the example with full_name btw)

Users have the original proto definitions and don't need type hints on the field. It is recommended to use a view extracting fields from a jsonb column for further analysis. Btw, using `any` type is a little tricky because it can involve multiple messages in one topic. I think the users realize it and we can keep the full name as a json field instead of introducing full names. 

> Why would you not to try keeping the original order? In my perspective this is like the column order in a traditional table.

users will get each field by name, I think no order needed here.

Reply:
> It is recommended to use a view extracting fields from a jsonb column for further analysis.

Could you provide an example use case of this, and does this mean you suggest store sub-fields as plain text?

> I think the users realize it and we can keep the full name as a json field instead of introducing full names.

I don't understand this part, do you suggest store the `full_name` as the json field or the `name` of the attribute as the json field?

> users will get each field by name, I think no order needed here.

If this is expected, PTAL the below reply. (To keep package name or not)

> > Rw gets both messageA and messageB wrapped in messageC. Rw will extract the inner fields without telling it is messageA or messageB
> 
> I don't think we could just extract the inner fields, say, the types here (i.e., string, int...etc) for every ingested message from upstream. Since the user will at least need the message name to further utilize the field / data from the parsed `Jsonb` column (e.g., `select any_column -> 'key' from t1;`). In this case if we simply extract the types of every incoming any message, user could not figure out which message they extract the field from. (e.g., The two `string` from `messageA` and `messageB` in the above case) Thus, I think it's reasonable to at least keep the name of the message, which includes the message name and the corresponding field name, makes it unique through a specific package. The further discussion is mainly regarding whether to keep the package name as a global unique identifier, to prevent naming conflict of messages from different packages.